pkg/reconciler/revision/reconcile_resources.go

```diff
@@ -113,18 +114,18 @@ func (c *Reconciler) reconcileImageCache(ctx context.Context, rev *v1.Revision)
 	logger := logging.FromContext(ctx)
 
 	ns := rev.Namespace
-	imageName := resourcenames.ImageCache(rev)
-	_, err := c.imageLister.Images(ns).Get(imageName)
-	if apierrs.IsNotFound(err) {
-		_, err := c.createImageCache(ctx, rev)
-		if err != nil {
-			return fmt.Errorf("failed to create image cache %q: %w", imageName, err)
+	for _, container := range rev.Status.ContainerStatuses {
```

Comment:
This will leak Image resources if I add/remove sidecars with unique names.

Reply:
In net-contour I devised a trick where I labeled children resource with the parent's generation, upserted the children that should exist and then did a DeleteCollection where generation was mismatched.

Reply:
> This will leak Image resources if I add/remove sidecars with unique names.

Hi @mattmoor I don't think i understand your point completely please help me to understand

How there will be a leak of image resources when names are unique :thinking:

Reply:
If a add a sidecar X with image Y and then replace it with sidecar names X with image Z, you won't replace the image?

Reply:
Replacing image will creates a new `Revision`
**Ex:**
Initially created a ksvc with 3 containers and the following are the image cache
```
kubectl get images.caching.internal.knative.dev
NAME                             AGE
multicontainer-v1-cache-first    4m16s
multicontainer-v1-cache-second   4m16s
multicontainer-v1-cache-third    4m16s
```
```
  name: multicontainer-v1-cache-second
  namespace: default
  ownerReferences:
  - apiVersion: serving.knative.dev/v1
    blockOwnerDeletion: true
    controller: true
    kind: Revision
    name: multicontainer-v1
    uid: 8507d109-3656-4f26-95e1-e30c1ee0ba9b
  resourceVersion: "275295"
  selfLink: /apis/caching.internal.knative.dev/v1alpha1/namespaces/default/images/multicontainer-v1-cache-second
  uid: 82f3f28b-9fff-43ee-86a3-770bced2cad4
spec:
  image: index.docker.io/savita3020/container-c@sha256:d8ccbc9b1bb48d7cdff5d41fb0b4ff2e634f8acf2abf665a334269369a95816a
```

Later updated the `ksvc` with the new image keeping name as you suggested
the latest image cache as follows
```
kubectl get images.caching.internal.knative.dev
NAME                             AGE
multicontainer-v1-cache-first    5m39s
multicontainer-v1-cache-second   5m39s
multicontainer-v1-cache-third    5m39s
multicontainer-v2-cache-first    16s
multicontainer-v2-cache-second   16s
multicontainer-v2-cache-third    16s
```
Replaced image for the container `second` keeping name as it is
```
  name: multicontainer-v2-cache-second
  namespace: default
  ownerReferences:
  - apiVersion: serving.knative.dev/v1
    blockOwnerDeletion: true
    controller: true
    kind: Revision
    name: multicontainer-v2
    uid: f4ababba-28a3-4c0e-a4c0-13dd6ad75e80
  resourceVersion: "275799"
  selfLink: /apis/caching.internal.knative.dev/v1alpha1/namespaces/default/images/multicontainer-v2-cache-second
  uid: 93b49909-ddae-47e4-99f1-9b1aad52ac46
spec:
  image: index.docker.io/savita3020/helloworld@sha256:90ef86ebfa7cc48e786bcc336258d0f3a9b225da644882eb5d057187c8ea37df
```

earlier we use to refer image cache with just `revisionName` now i have changed it to `revisionName+containerName`

Reply:
> If a add a sidecar X with image Y and then replace it with sidecar names X with image Z, you won't replace the image?

Yes replacing image will create a new image

A change to `ksvc` will creates a new `Revision`
**Ex:**
Initially created a ksvc with 3 containers and the following are the image cache
```
kubectl get images.caching.internal.knative.dev
NAME                             AGE
multicontainer-v1-cache-first    4m16s
multicontainer-v1-cache-second   4m16s
multicontainer-v1-cache-third    4m16s
```
```
  name: multicontainer-v1-cache-second
  namespace: default
  ownerReferences:
  - apiVersion: serving.knative.dev/v1
    blockOwnerDeletion: true
    controller: true
    kind: Revision
    name: multicontainer-v1
    uid: 8507d109-3656-4f26-95e1-e30c1ee0ba9b
  resourceVersion: "275295"
  selfLink: /apis/caching.internal.knative.dev/v1alpha1/namespaces/default/images/multicontainer-v1-cache-second
  uid: 82f3f28b-9fff-43ee-86a3-770bced2cad4
spec:
  image: index.docker.io/savita3020/container-c@sha256:d8ccbc9b1bb48d7cdff5d41fb0b4ff2e634f8acf2abf665a334269369a95816a
```

Later updated the `ksvc` with the new image keeping name as you suggested
the latest image cache as follows
```
kubectl get images.caching.internal.knative.dev
NAME                             AGE
multicontainer-v1-cache-first    5m39s
multicontainer-v1-cache-second   5m39s
multicontainer-v1-cache-third    5m39s
multicontainer-v2-cache-first    16s
multicontainer-v2-cache-second   16s
multicontainer-v2-cache-third    16s
```
Replaced image for the container `second` keeping name as it is
```
  name: multicontainer-v2-cache-second
  namespace: default
  ownerReferences:
  - apiVersion: serving.knative.dev/v1
    blockOwnerDeletion: true
    controller: true
    kind: Revision
    name: multicontainer-v2
    uid: f4ababba-28a3-4c0e-a4c0-13dd6ad75e80
  resourceVersion: "275799"
  selfLink: /apis/caching.internal.knative.dev/v1alpha1/namespaces/default/images/multicontainer-v2-cache-second
  uid: 93b49909-ddae-47e4-99f1-9b1aad52ac46
spec:
  image: index.docker.io/savita3020/helloworld@sha256:90ef86ebfa7cc48e786bcc336258d0f3a9b225da644882eb5d057187c8ea37df
```

earlier we use to refer image cache with just `revisionName` now i have changed it to `revisionName+containerName`

Reply:
Sorry, I just realized why what we have is fine, though we should probably document this in the code.  Revisions are immutable 🤦 

So we don't have to worry about updating digests or cleaning up ones that are no longer referenced because that set cannot change (without it being a new revision).  If you could add that clarification to the code it'd help future readers, but apologies for the false alarm.

Reply:
> Sorry, I just realized why what we have is fine, though we should probably document this in the code. Revisions are immutable 
> 
> So we don't have to worry about updating digests or cleaning up ones that are no longer referenced because that set cannot change (without it being a new revision). If you could add that clarification to the code it'd help future readers, but apologies for the false alarm.

No worries :) 
Actually without comment leads to this kind of confusion 
Thanks for pointing 

Updated with comment